### ISSUE TITLE
Questions related to temporal loss

### ISSUE THREAD

**Issue by zhengqili** [RaymondWang987/NVDS#20]
Hi, thanks for your great work for video depth!

I have a question related to temporal loss (Eq.4 in the main paper): in my opinion, the warped depth values from time i might not be equal to the corresponding depth value at time j if we don't take camera pose and 3D transformation into consideration. Do you think this is a bug or I miss something here?

**Comment by RaymondWang987** [comment]
> Hi, thanks for your great work for video depth!
> 
> I have a question related to temporal loss (Eq.4 in the main paper): in my opinion, the warped depth values from time i might not be equal to the corresponding depth value at time j if we don't take camera pose and 3D transformation into consideration. Do you think this is a bug or I miss something here?

Thank you for your approval and attention to our work. We have also discussed this question during the explorations and experimentations of our work.

Theoretically, it is a more reasonable way to use camera pose and geometric constraints to supervise the depth consistency in 3D space. Just like CVD, Robust-CVD, and Dynamic-Video-Depth, they project the corresponding points from frame i and j to 3D space and supervise the re-projection errors. However, in our case, we are faced with the problems below: 

(1) Our VDW training data is from natural-scene movies, documentaries, and web videos. It can be quite challenging to obtain accurate camera pose (R,t) and intrinsics K on such large scale in-the-wild video data. If the pose is inaccurate, the supervision with camera pose and 3D transformation will lead to depth errors and artifacts. This is also the main reason that leads to the artifacts and limited robustness of previous CVD, Robust-CVD, and Dynamic-Video-Depth in testing videos. They can only produce satisfactory results for the videos with accurate pose (from COLMAP as CVD or global pose optimization as Robust-CVD).

(2) The generalizability of the geometry projection losses is still questionable. For instance, the reprojection losses of CVD can only train the depth predictor to overfit the geometry constraints of a certain testing video, i.e., test-time training. The losses can be difficult to generalize to a large-scale dataset and supervise the model to learn the general consistency across different videos. In our work, we try to build a learning-based model that can directly predict consistent results, since the test-time training is very time-consuming.

To this end, we do not adapt these geometry reprojection errors with camera pose in 3D space (for robustness, generalizability, and efficiency). We use the optical-flow-based warping error following Cao et al. and FMNet. 

On the other hand, we also notice that the optical-flow-based warping error exists with some problems. Just like you said, the warped depth values from time i might not be equal to the corresponding depth value at time j. Our experimental results also prove this conclusion: the flow-based warping error and the spatial loss are actually conflicting to some extent. If the coefficient $\lambda$ of flow-based warping error is increased, it will cause blurring in the depth predictions. An excessively large $\lambda$ may even cause the model to fail to converge. Thus, we only use the temporal loss as an auxiliary loss with a small coefficient (e.g., $\lambda = 0.2$). The model itself can enforce consistency and adding the temporal loss will improve the temporal metric(as in Table 8 (a)). Besides, We have also taken some protective measures to mitigate this issue. For example, the temporal loss is calculated with the RGB masks as Cao et al., which means that pixels with larger RGB value changes will allow larger depth changes (i.e., lower weights in the temporal loss).

In my opinion, the geometry consistency losses with 3D transform are more reasonable than the flow-based loss. But the robust camera pose and the generalizability are challenging in our case. As a result, we just simply adopt the temporal loss as prior arts. Exploring more robust consistency supervision that complies with 3D geometric constraints is also a challenge that needs to be discussed in future works.
